### PR TITLE
Fixing the cookie_remap plugin to work with the new API

### PR DESCRIPTION
**Pull request by pushkar2017**

### CODE REVIEW

**Comment by SolidWallOfCode** [comment]
I think it's fine, I'd like @gtenev to take a look, because it's a change related to the remap plugin API change.

File: plugins/experimental/cookie_remap/cookie_remap.cc [hunk 1]
```diff
@@ -79,17 +79,24 @@ struct UrlComponents {
...
+    scheme    = TSUrlSchemeGet(bufp, urlp, &scheme_len);
+    host      = TSUrlHostGet(bufp, urlp, &host_len);
+    tspath    = TSUrlPathGet(bufp, urlp, &tspath_len);
+    query     = TSUrlHttpQueryGet(bufp, urlp, &query_len);
+    matrix    = TSUrlHttpParamsGet(bufp, urlp, &matrix_len);
+    port      = TSUrlPortGet(bufp, urlp);
+    url       = TSUrlStringGet(bufp, urlp, &url_len);
     from_path = TSUrlPathGet(rri->requestBufp, rri->mapFromUrl, &from_path_len);
 
+    TSHandleMLocRelease(bufp, TS_NULL_MLOC, urlp);
```
Comment (gtenev) [inline plugins/experimental/cookie_remap/cookie_remap.cc]: This seems odd to me. `TSHttpTxnPristineUrlGet()` returns a marshal buffer in `bufp` and location handle in `urlp`. Then `TSUrlSchemeGet()`, `TSUrlHostGet()`, etc. are called and the pointers pointing **inside** the marshal buffer are saved and used later **after** `TSHandleMLocRelease()` is called.

Although we don't do/use any reference counting currently and TSHandleMLocRelease() is a no-op for this type of objects (`HDR_HEAP_OBJ_URL`) I think we should make sure we don't release before we are done with the internal pointers.

Depending on the plugin needs we could (1) store the values pointed by the internal pointers and use them after the release or (2) keep the `bufp` and `urlp` until we are done with the pointers (for the lifetime of the `UrlComponents` class), similarly to cachekey plugin ([class CacheKey](https://github.com/apache/trafficserver/blob/master/plugins/cachekey/cachekey.cc#L249)) or tscppapi ([class ClientRequest](https://github.com/apache/trafficserver/blob/master/src/tscpp/api/ClientRequest.cc#L47)).

Comment (pushkar2017) [reply plugins/experimental/cookie_remap/cookie_remap.cc]: Good catch. Will fix it.

Comment (SolidWallOfCode) [reply plugins/experimental/cookie_remap/cookie_remap.cc]: I would be _very_ cautious about keeping the raw internal pointers, as those can go stale during any header access call. It is the `TSMLoc` that's stable, AFAICT that's the whole point of that type.

Comment (gtenev) [reply plugins/experimental/cookie_remap/cookie_remap.cc]: Good point! In fact this is what the cachekey and tscppapi do in the examples provided above.

**Comment by pushkar2017** [comment]
@gtenev please review. I release the handles in the destructor now. 
E.g.
TSHttpTxnPristineUrlGet(....)
TSUrlSchemeGet
...
Use raw pointers returned by the TSUrl APIs.
Release the handles acquired in TSHttpTxnPristineUrlGet.

I still don't understand how the raw pointers could be valid by just having a valid TSMBuffer and TSMLoc. Like you said, I don't see any ref counts or anything being updated in the implementation.

**Comment by SolidWallOfCode** [comment]
That's not how it works. The raw pointers are unstable, in that they can go stale during any API call that can cause a change in the header or URL object. The point of keeping the `TSBuffer` and `TSMLoc` is those are stable and from them the plugin can always get a valid raw pointer for immediate use. It would be reasonable to think of the raw pointers as iterators for a container like `std::vector`, which can be invalidated by any change in the container.

**Comment by gtenev** [comment]
@pushkar2017, [class CacheKey](https://github.com/apache/trafficserver/blob/master/plugins/cachekey/cachekey.cc#L249) is pretty similar to your case. It acquires + saves `TSBuffer` and `TSMloc`  handles during construction and releases them on destruction. In between for the lifetime of instance each member-function gets the internal pointers through the TS API right where they are immediately used. They are never cached since they can change if the object handled through `TSBuffer`+`TSMloc` gets modified.

**Comment by pushkar2017** [comment]
@gtenev Can you approve the PR now?

**Comment by gtenev** [comment]
@pushkar2017, thanks for fixing the plugin and enabling back its autest.

**Comment by shinrich** [comment]
[approve ci]

**Comment by zwoop** [comment]
[approve ci].

**Comment by randall** [comment]
[approve ci]

**Comment by bryancall** [comment]
Fixed conflict from clang-tidy on GitHub.

**Comment by bryancall** [comment]
[approve ci]

**Comment by randall** [comment]
[approve ci]

File: plugins/experimental/cookie_remap/cookie_remap.cc [hunk 2]
```diff
@@ -1059,8 +1080,8 @@ TSRemapDoRemap(void *ih, TSHttpTxn txnp, TSRemapRequestInfo *rri)
 
   for (auto &op : *ops) {
     TSDebug(MY_NAME, ">>> processing new operation");
-    if (op->process(jar, rewrite_to, status, rri)) {
-      cr_substitutions(rewrite_to, rri);
+    if ((*it)->process(jar, rewrite_to, status, rri, txnp)) {
```
Comment (randall) [inline plugins/experimental/cookie_remap/cookie_remap.cc]: Build failure here:
```
experimental/cookie_remap/cookie_remap.cc:1083:11: error: 'it' was not declared in this scope
     if ((*it)->process(jar, rewrite_to, status, rri, txnp)) {
           ^~
experimental/cookie_remap/cookie_remap.cc:1083:11: note: suggested alternative: 'ih'
     if ((*it)->process(jar, rewrite_to, status, rri, txnp)) {
           ^~
           ih
experimental/cookie_remap/cookie_remap.cc:1081:14: error: unused variable 'op' [-Werror=unused-variable]
   for (auto &op : *ops) {
              ^~
```

**Comment by bryancall** [comment]
@ywkaras is going to take a look at this.

**Comment by ywkaras** [comment]
[approve ci]

**Comment by ywkaras** [comment]
This PR has been replaced by https://github.com/apache/trafficserver/pull/7477 .

**Pull request closed without merging**